### ISSUE TITLE
how to specify different logging levels in drivers (and other extensions?)

### ISSUE THREAD

**Issue by matthewwall** [weewx/weewx#541]
Do we have a preferred approach for obtaining debug output from complex services/drivers? I am thinking here where we are getting byte streams that need to be decoded, often by reverse engineering rather than anything documented, so potentially a lot of cryptic output. I know that the implementation of the logging module under v4 gives us finer control, so we can limit debug output to just a single module. I presume this could also then use debug = 1,2,3 etc to progressively provide more debug output. Some drivers have individual debug output controls to say output debug info for just one particular aspect of the driver. And of course there is the
old way of simply using the debug option but that operates WeeWX wide and can provide a lot of output. I guess we need an approach that keeps it simple for the user to turn on/off (adding several config levels/options to a config file would seem risky, whilst preferably still limiting the output to that the the troubleshooter needs. Would something like the following work in weewx.conf:
```
....
debug = 0|1|2|3
....

[DriverX]
   ....
   debug_post = 0|1
   debug_decode = 0|1
   debug_crc = 0|1
   ....
```
where once the user turns on any of the normally hidden debug options in the driver the driver automatically limits debug output to the driver module only and the debug option is used to set the level. Sort of a two phase approach but keeps it simple for the user.

I am writing a driver and about to v4ise/ptython3ise another and and thought I might try and get things right from the start.

Gary

**Comment by matthewwall** [comment]
Good that you brought this up. 

Your proposal seems pretty reasonable. Off the top of my head, here are some other options that I can think of.

1. One of the nice things about 'debug=1' is that it requires no additional programming in the modules. You just use log.debug() and the system takes care of whether to emit or not. We could do something similar by adding, say, log.verbose(). Of course, that would apply system-wide, so the user would have to include a tailored [Logging] stanza to tune to what you want.

2. Use environment variables. Adding thought #1, and it might even be possible to modify the logger so that it would recognize named modules and turn on debugging just for that. Something like

WEEWX_vantage_DEBUG=2 weewxd

Pro: certainly powerful. Con: Complicated.

3. A hierarchical approach, like what log_success/log_failure uses. You could set debug=2 in your module, overriding values closer to the root. This would require considerable programming effort for every module that uses it, but is conceptually simple, with an existing pattern to follow.

Would be nice to accommodate Matthew's issue #371 as well.

I think of all these, I like your proposal the best. It's a little ad hoc, but very simple to implement and understand.

-tk

**Comment by matthewwall** [comment]
> 1. One of the nice things about 'debug=1' is that it requires no additional programming in the modules. You just use log.debug() and the system takes care of whether to emit or not. We could do something similar by adding, say, log.verbose(). Of course, that would apply system-wide, so the user would have to include a tailored [Logging] stanza to tune to what you want.

i like the simple syntax that is already in weewx:  just one flag, 'debug', with different levels, 0/1/2/3

although 'debug' is a misnomer, lets stick to just one word.  if you have 'debug' and 'verbosity', it is hard to tell which does what.

> 2. Use environment variables. Adding thought #1, and it might even be possible to modify the logger so that it would recognize named modules and turn on debugging just for that. 

please no environment variables!  java has this, and although it is powerful, it is cryptic.

lets keep everything in the config file


> 3. A hierarchical approach, like what log_success/log_failure uses. You could set debug=2 in your module, overriding values closer to the root. This would require considerable programming effort for every module that uses it, but is conceptually simple, with an existing pattern to follow.

we probably need a combination of this and the per-driver debug_* flags (see below for some caveats)


> Would something like the following work in weewx.conf:
>
>....
>debug = 0|1|2|3
>....
>
>[DriverX]
>    ....
>    debug_post = 0|1
>    debug_decode = 0|1
>    debug_crc = 0|1
>    ....

this is the pattern i have used in drivers.  the downside is that the driver options (debug_*) are not enabled unless the global debug is enabled.  when you only want driver info, it sucks to get all of the other info.  on the other hand, sometimes you *want* all the other stuff so you can see where the driver output fits relative to the rest.

with the additional flexibility of python logging, perhaps the driver options could be set independently of the global option.

btw, i think i used 0|1|2 in some cases of driver debug.  for example, 1 gives you raw output, 2 gives you raw output plus decoded output.

basically the flag name (debug_io, debug_writes, etc) are the mechanism for saying "these are the things you can report about" while the 0/1/2/... is the mechanism for saying "i want more detail about this thing"

m
